File: LeetCode/785. 判断二分图.ts
function isBipartite(graph: number[][]): boolean {
    let valid=true;
    let n = graph.length;
    let colors=Array(n).fill(0);
    function dfs(i:number,color:number){
        colors[i]=color;
        let nextColor=-color;
        for(let neighbor of graph[i]){
            if(colors[neighbor]===0){
                dfs(neighbor,nextColor);
            }else if(colors[neighbor]!==nextColor){
                valid=false;
                return;
            }
        }
    }
    for(let i=0;i<n&&valid;i++){
        if(colors[i]===0){
            dfs(i,-1);
        }
    }
    return valid;
};

File: LeetCode/785. 判断二分图.cpp
class Solution {
private:
    bool valid;
    vector<int> colors;
public:
    bool isBipartite(vector<vector<int>>& graph) {
        valid=true;
        int n=graph.size();
        colors.assign(n,0);
        for(int i=0;i<n&&valid;i++){
            if(colors[i]==0){
                dfs(i,-1,graph);
            }
        }
        return valid;
    }
    void dfs(int index,int color,vector<vector<int>>&graph){
        colors[index]=color;
        int nextColor=-color;
        for(int neighbor:graph[index]){
            if(colors[neighbor]==0){
                dfs(neighbor,nextColor,graph);
            }else if(colors[neighbor]!=nextColor){
                valid=false;
                return;
            }
        }
    }
};
